Евгений Петров

LeMarck opened the pull request.

LeMarck committed:
Add code

bratva commented on phoneBook.js:
```diff
@@ -1,52 +1,117 @@
+        return false;
+    if (!emailValidator.test(email))
+        return false;
+    maxSize[0] = (name.length > maxSize[0] ? name.length : maxSize[0]);
```
Используй `Math.max(name.length, maxSize[0])`

bratva commented on phoneBook.js:
```diff
@@ -1,52 +1,117 @@
+    var searchItem = [];
+    for (var index = 0; index < phoneBook.length; index++)
+        if (searchForm.test(phoneBook[index].name) || searchForm.test(phoneBook[index].phone) || searchForm.test(phoneBook[index].email))
+            searchItem.push( phoneBook[index]);
```
Лишний пробел

bratva commented on phoneBook.js:
```diff
@@ -1,52 +1,117 @@
 
     // Ваша удивительная магия здесь
 
+    var item = search(query);
```
`search` у тебя возвращаем массив элементов. Переменную лучше назвать `items`;

bratva commented on phoneBook.js:
```diff
@@ -1,52 +1,117 @@
 
+    var item = search(query);
+    for (var index = 0; index < item.length; index++)
+        console.log(item[index].name + ', ' + item[index].phone + ', ' + item[index].email);
```
Номера телефонов должны выводиться нормализованными, в виде `+7 (999) 666-7-778`. А не так как их записали в базу.

bratva commented on phoneBook.js:
```diff
@@ -1,52 +1,117 @@
 
     // Ваша необьяснимая магия здесь
 
+    var item = search(query);
```
`items`.
Молодец что унес поиск элементов в отдельную фукцию

bratva commented on phoneBook.js:
```diff
@@ -1,52 +1,117 @@
 
+    var item = search(query);
+    for (var index=0; index < item.length; index++)
+        phoneBook.pop(item[index]);
```
Нашел одни элементы, а удаляешь другие (`pop` - достает последний элемент их стека)

bratva commented:
привет @LeMarck 

Сейчас программа работает не верно. Необходимо починить.
Так же поправь указанные замечания и еще раз проверь на [codestyle](https://github.com/urfu-2015/guides/blob/master/js-codestyle.md)

LeMarck committed:
Corrected code

bratva commented:
Поправь программу. Сейчас есть проблемы:
- Не добавляется телефон +7 (111) 777 2 222 - хотя он валиден
- Почему-то ты посчитал телефон вида +7 (111 777-2-222, 7+ (111) 777-2-222, -7 (111) 777-2-222 - валидным.
- Электропочта info@yandex - не валидна, но ты добавил ее
- Не получается найти телефон по выражению вида '+7 (999)'
- Не и удаление собственно такого запроса не происходит

bratva commented:
:star: - Молодец, что выполнил доп. часть.

LeMarck committed:
Fixed

bratva commented:
Программа падает с ошибкой 

``` js
TypeError: Cannot read property 'replace' of undefined // строка 37
```

Происходит это при вызове `.find()` без аргументов

bratva commented:
Емейл `info@инфо.рф` - вполне валиден и допустим - но увы с ним не работает

LeMarck committed:
Corrected code

@bratva, добавил проверки на пустоту не только в find, но и в другие функции.
Совсем забыл, что \w - это только латиница, добавил в регулярку \W и всё стало ОК

bratva commented on phoneBook.js:
```diff
@@ -1,52 +1,155 @@
 
     // Ваша удивительная магия здесь
-
+    if (query){
```
Если query не задан (`undefined`) - то нужно выводить все записи.

bratva commented on phoneBook.js:
```diff
@@ -1,52 +1,155 @@
+    }
+    var phoneValidator = /^\+?(\d{0,3})\s?(\(\d{3}\)|\d{3})\s?(\d{3})[- ]?(\d)[- ]?(\d{3})$/;
+    var codeNumberValidator = /^\(?(\d{3})\)?$/;
+    var emailValidator = /^[\W\w]+@([\W\w-]+\.)+[\W\w]+$/;
```
``` js
/^[\W\w]+@([\W\w-]+\.)+[\W\w]+$/.test('info@yandex@ya.ru'); // true
```

bratva commented on phoneBook.js:
```diff
@@ -1,52 +1,155 @@
+ Функция добавления записи в телефонную книгу.
+ На вход может прийти что угодно, будьте осторожны.
+ */
 module.exports.add = function add(name, phone, email) {
```
Сейчас функция добавления вылядит грязной. В ней делается слишком многое:
- Валидация
- Подсчет максимальной длины
- Форматирование
- Добавление (В идеале надо оставить только это)

Может попробуешь вынести эти куски кода в другие функции?
Например 

``` js
function isPhoneValid (phone) {
    ...
    return true / false;
}

function updateMaxSize (name, phone, email) {

}

...
```

А тут сделаешь так

``` js
module.exports.add = function add(name, phone, email) {
   ...
   if (isPhoneValid(phone)) {
      ...
   }
}
```

bratva commented:
Женя - осталось еще пару проблем. Почини и зачту. В целом молодец.

LeMarck committed:
Added fixes

@bratva, поправил код. Регулярка для емайл вроде должна нормально работать

bratva commented:
Хорошо
